keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,435 @@
+Generators generate 0 or more Resources for some Virtual Resources.
+
+Example: Kustomize secretGenerators and configMapGenerators generate Secrets and ConfigMaps
+from a `kustomize.config.k8s.io/v1beta1/Kustomization` virtual Resource.
+
+Generators have 2 components:
+
+1. Generator Config
+  - Virtual Resource
+  - Added to the `kustomization.yaml` field `generators`
+1. Generator Implementation
+  - Executable plugin
+  - Reads Virtual Resources
+  - Emits non-Virtual Resources
+
+1. Kustomize reads the `generators` Virtual Resources
+1. Kustomize maps the Virtual Resources to plugins by their *Group*
+   - Group matches the plugin name
+   - Exits non-0 if no plugins are found any generator entries
+1. For each Virtual Resource *Group*
+  1. Kustomize execs the plugin `generate` command
+  1. Kustomize writes the Virtual Resources in that Group to the process STDIN
+  1. Kustomize reads the set of generated Resources from the process STDOUT
+  1. Kustomize reads error messages from the exec process STDERR
+  1. Kustomize fails if the plugin exits non-0
+  1. Kustomize adds the emitted Resources to its set of Non-Virtual Resources (e.g. from `resources`, `bases`, etc).
+
+The order of plugin execution is arbitrary.
+
+#### Plugins: Transform
+
+Transformers modify existing non-virtual Resources by modifying their fields.
+
+Transformers have 2 components:
+
+1. Transformer Config
+  - Virtual Resource
+  - Added to the `kustomization.yaml` field `transformers`
+  - All `generators` are implicitly invoked as transformers
+1. Transformer Implementation
+  - Executable plugin
+  - Reads Virtual Resources
+  - Reads **all** non-Virtual Resources
+  - Emits non-Virtual Resources  
+  
+1. Kustomize reads the `transformers` *and* `generators` Virtual Resources
+  - `generators` can require transformation
+1. Kustomize maps the Virtual Resources to plugins by their *Group*
+   - Group matches the plugin name
+   - Exits non-0 if no plugins are found any generator entries
+1. For each Virtual Resource *Group*
+  1. Kustomize execs the plugin `transform` command
+  1. Kustomize writes the Virtual Resources in that Group to the process STDIN
+  1. Kustomize writes **all** non-Virtual Resources to the process STDIN
+  1. Kustomize reads the set of transformed Resources from the process STDOUT
+  1. Kustomize reads error messages from the exec process STDERR
+  1. Kustomize fails if the plugin exits non-0
+  1. Kustomize replaces its current set of non-Virtual Resources with the set of emitted Resources.
+
+The order of plugin execution is arbitrary.
+
+##### Restrictions
+
+For the initial iteration, transformers cannot add/remove Resources, or change their names/namespaces.
+The ability doing so would be significantly make complex ordering interactions much more likely.  E.g. transformers
+would need to keep and propagate transformations from other transformers.
+
+We are prioritizing a restrictive but predictable API over a powerful but complex one.
```

Comment:
(I've already mentioned it several times before, so this will be my last:  You can't usefully use multiple transformers without being able to control ordering.  Non-commutative transformations are definitely not limited to just changing name/namespace and the proposed "arbitrary ordering" does not lead to predictable output.

I understand that you really don't want to have any ordered lists for some reason, so I'll just keep this passive-aggressive "I told you so" comment here to point back to in the future :stuck_out_tongue: )

Reply:
+1.  FWIW, I am not against this, I just don't have enough data to understand the requires - e.g.  concrete expected use cases vs conceptual academic use cases.

When this feature has enough adoption that we have this data, I will look for a proposal that uses them to define requirements and come up with a design.

Since supporting explicit ordering is backwards compatible with arbitrary ordering - it is simpler to add in the future, and I am comfortable deferring the decision the specific API that would support this - e.g. is it:
- an ordered list of transformer plugins only
- an ordered list of transformer plugins and generator plugins only
- an absolute ordering of both built-in and plugin transformers and generators
- an acyclic dependency graph
- an cyclic dependency graph
- additional transformation phases
- automatic ordering based on transformer duck-typing or interfaces
- some combination of the above